### PR TITLE
Dev kubernetesconfiguration microsoft.kubernetes configuration 2022 01 15 preview

### PR DESCRIPTION
**Pull request by deeksha345**
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [x] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  4. If updating an existing version, please select the specific langauge SDKs and CLIs that must be refreshed after the swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [ ] No refresh required for updates in this PR

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist 

> **Applicability**: :warning: 
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs 
> - Adding new properties 
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [x] Check this box if any of the following apply to the PR so that label "WaitForARMFeedback" will be added automatically to begin ARM API Review. Failure to comply may result in delays to the manifest.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you are using OpenAPIHub to initialize the PR for adding a new version. More details, refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).

- [x] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If any of the following scenarios apply to the PR, request approval from the Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 
- [ ] Removing API(s) in a stable version
- [ ] Removing properties in a stable version
- [ ] Removing API version(s) in a stable version
- [ ] Updating API in a stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

**Comment by deeksha345** [comment]
@ArcturusZhang 

For Cross-Version Breaking Changes

1. /subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/{clusterRp}/{clusterResourceName}/{clusterName}/providers/Microsoft.KubernetesConfiguration/operations path still exists, but has been moved to operations.json, this is not breaking
2. providers/Microsoft.KubernetesConfiguration/operations path still exists, but has been moved to operations.json, this is not breaking
3. ReadOnlyProperty changes in kubernetesconfiguration.json have already been approved in previous API reviews for 2021-05-01-preview and 2021-06-01-preview, these should not be considered breaking

File: specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/common/2022-01-15-preview/parameters.json [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+      "description": "Azure Active Directory OAuth2 Flow",
+      "scopes": {
+        "user_impersonation": "impersonate your user account"
+      }
+    }
+  },
+  "paths": {},
+  "definitions": {},
+  "parameters": {
+    "ClusterRpParameter": {
```
Comment (mentat9) [inline specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/common/2022-01-15-preview/parameters.json:29]: >ClusterRpParameter [](http://example.com/codeflow?start=5&length=18)

This parameter had its x-ms-enum extension removed (previously had modelAsString: true). This removal isn't an ARM issue, but might be an SDK issue.

File: specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/common/2022-01-15-preview/parameters.json [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+      "required": true,
+      "type": "string",
+      "enum": [
+        "Microsoft.ContainerService",
+        "Microsoft.Kubernetes"
+      ],
+      "description": "The Kubernetes cluster RP - either Microsoft.ContainerService (for AKS clusters) or Microsoft.Kubernetes (for OnPrem K8S clusters).",
+      "x-ms-parameter-location": "method"
+    },
+    "ClusterResourceNameParameter": {
```
Comment (mentat9) [inline specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/common/2022-01-15-preview/parameters.json:41]: >ClusterResourceNameParameter [](http://example.com/codeflow?start=5&length=28)

Same thing here: x-mx-enum was removed w.r.t. the previous API version (2022-01-01-preview).

File: specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/preview/2022-01-15-preview/extensionTypes.json [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+          "type": "array",
+          "items": {
+            "type": "string"
+          },
+          "description": "Extension release train: preview or stable"
+        },
+        "clusterTypes": {
+          "description": "Cluster types",
+          "readOnly": true,
+          "type": "array",
```
Comment (mentat9) [inline specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/preview/2022-01-15-preview/extensionTypes.json:275]: >array [](http://example.com/codeflow?start=19&length=5)

This is a cross-version breaking change. You will need to get a breaking change review.

File: specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/preview/2022-01-15-preview/extensionTypes.json [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+        "defaultReleaseNamespace": {
+          "type": "string",
+          "description": "Default extension release namespace"
+        }
+      }
+    },
+    "NamespaceScopeSettings": {
+      "type": "object",
+      "description": "Extension scope settings",
+      "properties": {}
```
Comment (mentat9) [inline specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/preview/2022-01-15-preview/extensionTypes.json:358]: >properties": {} [](http://example.com/codeflow?start=7&length=15)

ARM recommends against the use of the additionalProperties/empty/unschematized object pattern. ARM favors stronger typing wherever possible because it allows all consumers of your API to reason about it. This has benefits for everyone that interacts with your API including basic infrastructure, other Azure features, SDK owners, and customers.

For example, customers will not be able to manage these objects using Azure Policy, which is a standard ARM feature. Your customers may be unhappy about that. There are other features like ARG and others under development that won't be able to support these objects. Unschematized objects are essentially API contracts that bypass Azure versioning infrastructure.

Generally, this pattern should only be used when needed and the object properties are dynamic, unknown, or user defined. To address this you can either enumerate the object's properties in your swagger, implement it as its own resource type, or remove the object from the spec.

Ref: [https://armwiki.azurewebsites.net/api_contracts/guidelines/openapi.html#oapi032-only-use-additionalproperties-when-the-object-properties-are-dynamic-unknown-or-user-defined](https://armwiki.azurewebsites.net/api_contracts/guidelines/openapi.html#oapi032-only-use-additionalproperties-when-the-object-properties-are-dynamic-unknown-or-user-defined).

File: specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/common/2022-01-15-preview/parameters.json (hunk 1 above)
Comment (deeksha345) [reply specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/common/2022-01-15-preview/parameters.json:29]: I am not seeing any difference between my parameters and the ones from previous versions. Can you provide more detail on what this is referring to?

File: specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/common/2022-01-15-preview/parameters.json (hunk 2 above)
Comment (deeksha345) [reply specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/common/2022-01-15-preview/parameters.json:41]: I am not seeing any difference between my parameters and the ones from previous version 2022-01-01. Can you provide more detail on what this is referring to?

Comment (mentat9) [reply specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/common/2022-01-15-preview/parameters.json:41]: The two parameter definitions in parameters.json had this removed:
```
      "x-ms-enum": {
        "name": "ExtensionsClusterRp",
        "modelAsString": true
      },
```
from ClusterRpParameter and this
```
      "x-ms-enum": {
        "name": "ExtensionsClusterResourceName",
        "modelAsString": true
      },
```
from ClusterResourceNameParameter. As I said, I don't think it affects the REST layer, but it affects SDK generation.

**Comment by mentat9** [comment]
@deeksha345, please add the WaitForARMFeedback or WaitForARMRevisit label when you are ready for revisit: that will notify the current oncall that your PR is ready for revisit.

File: specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/common/2022-01-15-preview/parameters.json (hunk 2 above)
Comment (deeksha345) [reply specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/common/2022-01-15-preview/parameters.json:41]: Ok I see there were some changes made after I created this fork. I will update these properties

File: specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/preview/2022-01-15-preview/extensionTypes.json (hunk 3 above)
Comment (deeksha345) [reply specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/preview/2022-01-15-preview/extensionTypes.json:275]: So this service was never released to production before this point. The ARM Manifest has still not been updated with these new resource types. Is breaking change review still required?

File: specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/preview/2022-01-15-preview/extensionTypes.json (hunk 4 above)
Comment (deeksha345) [reply specification/kubernetesconfiguration/resource-manager/Microsoft.KubernetesConfiguration/preview/2022-01-15-preview/extensionTypes.json:358]: Good point I will remove this property.

**Comment by deeksha345** [comment]
@mentat9 
Can you please revisit this PR, I have made all the necessary updates based on your previous comments.

**Comment by mentat9** [comment]
> @mentat9 Can you please revisit this PR, I have made all the necessary updates based on your previous comments.

@deeksha345, please work with current oncall (@rkmanda) to get your PR ARM-reviewed.

Here are some general tips to help with ARM review:
1. You can see the ARM review rotation at IcM -> On Call Lists -> Who's On Call -> Azure Resource Manager / RP Manifest Approvers.
2. You can always ping the current oncall for help with ARM review of your PR. They are very busy with incoming requests but will always try to help.
3. ARM review oncall changes every week, so if you don't get a review the end of the week, it's a good idea to ping the new reviewer on Monday to get in their queue.

**Comment by ArcturusZhang** [comment]
Hi @deeksha345 please also submit a breaking change review request as well since your PR has some cross-version breaking changes. You can find the guidance here: https://github.com/Azure/azure-rest-api-specs/pull/17516#issuecomment-1022674515

**Comment by deeksha345** [comment]
@jorgecotillo Can you please review this PR

**Comment by deeksha345** [comment]
@jorgecotillo 

For Cross-Version Breaking Changes

/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/{clusterRp}/{clusterResourceName}/{clusterName}/providers/Microsoft.KubernetesConfiguration/operations path still exists, but has been moved to operations.json, this is not breaking
providers/Microsoft.KubernetesConfiguration/operations path still exists, but has been moved to operations.json, this is not breaking
ReadOnlyProperty changes in kubernetesconfiguration.json have already been approved in previous API reviews for 2021-05-01-preview and 2021-06-01-preview, these should not be considered breaking

**Review by j5lim — APPROVED** [review]
Approved from ARM side.

**Comment by deeksha345** [comment]
@ArcturusZhang Can you please approve and merge these changes now that ARM has signed off.

**Comment by deeksha345** [comment]
@ArcturusZhang Yes please do ahead and update the default tag.

**Comment by ArcturusZhang** [comment]
Hi @deeksha345 we have quite a few conflicts in this PR right now, could you please resolve those? And please also change the default tag to the new tag added in this PR

File: specification/kubernetesconfiguration/resource-manager/readme.md [hunk 5]
```diff
@@ -28,12 +28,10 @@ These are the global settings for the KubernetesConfiguration.
 title: SourceControlConfigurationClient
 description: KubernetesConfiguration Client
 openapi-type: arm
-tag: package-2022-03
+tag: package-preview-2022-01-15
```
Comment (ArcturusZhang) [inline specification/kubernetesconfiguration/resource-manager/readme.md:31]: Let us do not downgrade the tag. Could you please revert this change?
Our SDK releasing process is depending on this tag, downgrade this will make the SDK to downgrade by api-versions

Comment (deeksha345) [reply specification/kubernetesconfiguration/resource-manager/readme.md:31]: Ok I will revert this change. 

Can you please let me know what I need to change regarding this comment you made previously then: "Please also change the default tag to the new tag added in this PR."
I am not sure what the tag needs to be changed to then.

**Comment by ArcturusZhang** [comment]
The cross version breaking changes come from the comparison with the latest stable version which could be ignored

**Comment by ruowan** [comment]
Pls fix avocado error. Add your unreferenced swagger file into readme file.

**Review by ArcturusZhang — APPROVED** [review]
